### PR TITLE
feat: Add SNAT bridge to Native, decouple SNAT bridge

### PR DESCRIPTION
**Pull request by QxBytes**
**Reason for Change**:
Adds SNAT bridge functionality to the native datapath and decouples the OVS-specific SNAT bridge code from the more general SNAT bridge code.

**Issue Fixed**:
N/A

**Requirements**:

- [X] uses [conventional commit messages](https://www.conventionalcommits.org/)
- [X] includes documentation
- [X] adds unit tests

**Notes**:
- The network package was `gofumpt` so there may be style fixes in the following files not directly related to this feature (See commit "Go fumpt entire network package" for those files). 
- Some code was moved around/renamed/re-used. In particular, snat_linux.go (and most other files) consists of already-existing code that was re-used.
- The idea is that OVS and Native have their own versions of SNAT/snat route, which call a common set of behaviors in endpoint_snatroute_linux.go, which uses the client to call something in snat_linux.go

### CODE REVIEW

**Comment by rbtr** [comment]
could you pull the gofumpt of the whole package out into a separate PR? happy to merge that one first

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]
i see you updates ovs snat as well.. did you test snat bridge scenarios work in ovs case?

File: cni/azure-windows-multitenancy-transparent-vlan.conflist [hunk 1]
```diff
@@ -0,0 +1,52 @@
+{
+    "cniVersion": "0.3.0",
+    "name": "azure",
+    "plugins": [
+        {
+            "type": "azure-vnet",
+            "mode": "transparentVlan",
```
Comment (tamilmani1989) [inline cni/azure-windows-multitenancy-transparent-vlan.conflist:7]: transparent-vlan

File: network/endpoint_snatroute_linux.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		return errors.Wrap(err, "failed to enable ip forwarding")
+	}
+
+	if hostToNC {
+		if err := snatClient.AllowInboundFromHostToNC(); err != nil {
+			return errors.Wrap(err, "failed to allow inbound from host to nc")
+		}
+	}
+
+	if ncToHost {
```
Comment (tamilmani1989) [inline network/endpoint_snatroute_linux.go:47]: snatClient.AllowInboundNCToHost()

File: network/endpoint_snatroute_linux.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		return errors.Wrap(snatClient.AllowInboundFromNCToHost(), "failed to allow inbound from nc to host")
+	}
+	return nil
+}
+
+func MoveSnatEndpointToContainerNS(snatClient *snat.Client, netnsPath string, nsID uintptr) error {
+	return errors.Wrap(snatClient.MoveSnatEndpointToContainerNS(netnsPath, nsID), "failed to move snat endpoint to container ns")
```
Comment (tamilmani1989) [inline network/endpoint_snatroute_linux.go:54]: change everything to same format

File: network/network.go [hunk 4]
```diff
@@ -15,11 +15,11 @@ import (
...
 	// Operational modes.
-	opModeBridge      = "bridge"
-	opModeTunnel      = "tunnel"
-	opModeTransparent = "transparent"
-	opModeNative      = "native"
-	opModeDefault     = opModeTunnel
+	opModeBridge          = "bridge"
+	opModeTunnel          = "tunnel"
+	opModeTransparent     = "transparent"
+	opModeTransparentVlan = "transparentVlan"
```
Comment (tamilmani1989) [inline network/network.go:21]: transparent-vlan

File: network/ovs_endpoint_snatroute_linux.go [hunk 5]
```diff
@@ -1,114 +1,151 @@
...
+	}
 	return nil
 }
 
-func MoveSnatEndpointToContainerNS(client *OVSEndpointClient, netnsPath string, nsID uintptr) error {
-	if client.enableSnatOnHost || client.allowInboundFromHostToNC || client.allowInboundFromNCToHost || client.enableSnatForDns {
-		return client.snatClient.MoveSnatEndpointToContainerNS(netnsPath, nsID)
+func (client *OVSEndpointClient) AddSnatEndpointRules() error {
+	if client.isSnatEnabled() {
+		// Add route for 169.254.169.54 in host via azure0, otherwise it will route via snat bridge
```
Comment (tamilmani1989) [inline network/ovs_endpoint_snatroute_linux.go:105]: 169.254.169.**2**54

File: network/transparent_vlan_endpointclient_linux.go [hunk 6]
```diff
@@ -109,15 +156,22 @@ func (client *NativeEndpointClient) PopulateVM(epInfo *EndpointInfo) error {
...
 		// Create vlan veth
 		deleteNSIfNotNilErr = vishnetlink.LinkAdd(link)
 		if deleteNSIfNotNilErr != nil {
 			// Any failure to add the link should error (auto delete NS)
 			return errors.Wrap(deleteNSIfNotNilErr, "failed to create vlan vnet link after making new ns")
 		}
+		deleteNSIfNotNilErr = client.netUtilsClient.DisableRAForInterface(client.vlanIfName)
+		if deleteNSIfNotNilErr != nil {
+			if delErr := client.netlink.DeleteLink(client.vlanIfName); delErr != nil {
+				log.Errorf("deleting vlan veth failed on addendpoint failure")
```
Comment (tamilmani1989) [inline network/transparent_vlan_endpointclient_linux.go:169]: move to defer

File: network/transparent_vlan_endpointclient_linux.go [hunk 7]
```diff
@@ -163,49 +230,59 @@ func (client *NativeEndpointClient) PopulateVnet(epInfo *EndpointInfo) error {
...
 // Adds routes, arp entries, etc. to the vnet and container namespaces
-func (client *NativeEndpointClient) ConfigureContainerInterfacesAndRoutes(epInfo *EndpointInfo) error {
+func (client *TransparentVlanEndpointClient) ConfigureContainerInterfacesAndRoutes(epInfo *EndpointInfo) error {
 	// Container NS
 	err := client.ConfigureContainerInterfacesAndRoutesImpl(epInfo)
 	if err != nil {
 		return err
 	}
 
+	if err = client.ConfigureSnatContainerInterface(); err != nil {
```
Comment (tamilmani1989) [inline network/transparent_vlan_endpointclient_linux.go:274]: move this to end of function

File: network/transparent_vlan_endpointclient_linux.go [hunk 8]
```diff
@@ -325,7 +402,11 @@ func (client *NativeEndpointClient) AddDefaultArp(interfaceName, destMac string)
 	return nil
 }
 
-func (client *NativeEndpointClient) DeleteEndpoints(ep *endpoint) error {
+func (client *TransparentVlanEndpointClient) DeleteEndpoints(ep *endpoint) error {
+	if err := client.DeleteSnatEndpoint(); err != nil {
+		return errors.Wrap(err, "failed to delete snat endpoint")
+	}
```
Comment (tamilmani1989) [inline network/transparent_vlan_endpointclient_linux.go:408]: move to end of function

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

**Review by tamilmani1989 — APPROVED** [review]

**Pull request merged**